Add cover image

georgeh opened the pull request:
Addresses #487 

# TODO

- [x] Remove text formatting options
- [x] Add full-height/parallax options
- [ ] ~Pre-select image in image chooser when replacing the image~

Commit:
Cover Image is showing up in the editor… kinda…

Commit:
background image not showing up

Commit:
Fixed background image style

mtias commented on post-content.js:
```diff
@@ -7,6 +7,14 @@ window._wpGutenbergPost = {
+			'<!-- /wp:core/cover-image -->',
+
+			'<!-- wp:core/heading -->',
+			'<h1>Welcome to the Gutenberg Editor</h1>',
```
This is no longer part of the demo content.

georgeh replied:
Thanks, must have been a merge conflict.

youknowriad commented on blocks/library/cover-image/index.js:
```diff
@@ -0,0 +1,155 @@
+					</div>
+					<Button isLarge>
+						{ wp.i18n.__( 'Insert from Media Library' ) }
+					</Button>
```
The media library integration is already available through the `MediaUploadButton`. see https://github.com/WordPress/gutenberg/blob/master/blocks/library/image/index.js#L89

mtias replied:
We need to start thinking about documenting these (editable, media button, the iframe thing).

youknowriad commented on blocks/library/cover-image/index.js:
```diff
@@ -0,0 +1,155 @@
+			title: wp.i18n.__( 'Change Image' ),
+			isActive: ( { align } ) => 'left' === align,
+			onClick: toggleAlignment( 'left' ),
+		},
```
What is this control for? should we drop it?

youknowriad commented on blocks/library/cover-image/index.js:
```diff
@@ -0,0 +1,155 @@
+	category: 'common',
+
+	attributes: {
+		url: attr( 'section', 'data-url' ),
```
FWIW, it's possible to drop the `url` attribute here and the url will be saved in the block opening comment.

youknowriad commented on blocks/library/cover-image/index.js:
```diff
@@ -0,0 +1,155 @@
+import { registerBlock, query } from '../../api';
+import Editable from '../../editable';
+// TODO: Revisit when we have a common components solution
+import Dashicon from '../../../components/dashicon';
```
Should we move this to the WordPress dependencies `import Dashicon from 'components/dashicon';`?

youknowriad reviewed (commented):
Thanks for your help on this block. I left some small notes

georgeh commented:
@mtias and @youknowriad thank you for the 👀 , the feedback is really helpful for me (as someone less familiar with the project). I'm still getting the block working, and after that I'll work on cleaning everything up.

Commit:
Clean up and remove outdated code

Commit:
Use MediaUploadButton

Commit:
Cover images working in the editor

Thanks to @mkaz for the pairing session to get it done

Commit:
registerBlock ➡️ registerBlockType

Commit:
Fix tests

Commit:
Remove formatting options for title

georgeh commented:
@jasmussen would you kindly try this block out and let me know how it feels? I'm still working on a parallax setting but the core functionality should be there

jasmussen commented:
I like it! I have a bunch of small things, including a different UI for adding/changing the image, but I think we might want to try and keep things minimal for the first merge, then iterate. 

There isn't any `background-attachment: fixed`, though —  any reason why? I think that'd be a nice feature to have.

georgeh commented:
@jasmussen I wanted to make sure I was on the right path before going too far down it. I'm going to play with a few settings for `background-attachment: fixed` to see what I like

Commit:
Add background-attachment: fixed view

Commit:
Pre-selecting image by URL doesn't make sense

georgeh commented:
@youknowriad could you take another look at this? I'm ready to merge

mkaz commented:
* Backspace doesn't work in title field

* I think I get the Pin for background fixed, but it seems to also align-right, should pin be separate then the alignments - ie. I can pin a full bleed image

georgeh commented:
@mkaz thanks, I'll check out the backspace. 

For alignment vs pin I see the argument for allowing something to be both aligned and pinned. I'll try out a few CSS tweaks to see how they look.

youknowriad commented on blocks/library/cover-image/index.js:
```diff
@@ -0,0 +1,169 @@
+/**
+ * WordPress dependencies
+ */
+import Placeholder from 'components/placeholder';
```
We should use `import { Placeholder } from 'components';` to avoid duplication in the built files. (We should probably make this clearer in our dev docs)

youknowriad commented on blocks/library/cover-image/index.js:
```diff
@@ -0,0 +1,169 @@
+		title: text( 'h2' ),
+	},
+
+	controls: [
```
If you rebase master, you'll notice this property doesn't work anymore and I suggest using `<BlockAlignmentToolbar>` for alignment controls. See #1019 for examples.

youknowriad commented on blocks/library/cover-image/index.js:
```diff
@@ -0,0 +1,169 @@
+};
+
+registerBlockType( 'core/cover-image', {
+	title: wp.i18n.__( 'Cover Image' ),
```
We're trying to avoid using the `wp` global variable as much as we can in favor of `import { __ } from 'i18n`. See https://github.com/WordPress/gutenberg/pull/1332 for context on this.

youknowriad commented on blocks/library/cover-image/index.js:
```diff
@@ -0,0 +1,169 @@
+
+	editFrame.on( 'insert', updateFn );
+	editFrame.open( 'cover-image' );
+};
```
This is the second block allowing to edit a media. I wonder if should try to factorize this in `MediaUploadButton` or similar. This could be done in a separate PR.

georgeh replied:
I'm going to remove this for now but I think it may make sense as a follow-up as it's own `InspectorControl`.

youknowriad commented on blocks/library/cover-image/index.js:
```diff
@@ -0,0 +1,169 @@
+						buttonProps={ uploadButtonProps }
+						onSelect={ setMediaUrl }
+						type="image"
+						auto-open
```
auto-open has been rename `autoOpen`

mtias commented:
@georgeh we should try to use `git rebase` instead of merges from master.

Commit:
Control updates

Now using the new `<BlockControls>` to manage the image position.

Also removed the ability to change the picture pending moving it into the inspector

Also addressed changes in the PR to remove `wp` global and fix autoOpen the MediaUploadButton

mtias commented on blocks/block-alignment-toolbar/index.js:
```diff
@@ -25,6 +25,10 @@ const BLOCK_ALIGNMENTS_CONTROLS = {
 		icon: 'align-full-width',
 		title: __( 'Full width' ),
 	},
+	fixed: {
```
I think this should be a toggle in the inspector.

georgeh replied:
I'm going to make a `CheckboxControl` to use in the inspector

mtias replied:
Should probably just use the existing toggle for this. cc @jasmussen

georgeh replied:
Looking at the code for `TextboxControl` it looks like I can pass `type="checkbox"` - how do you feel about me renaming that to `InputControl`?

jasmussen replied:
Yes if we were to make a toggle for this it should just be the existing switch:

![screen shot 2017-06-22 at 19 57 02](https://user-images.githubusercontent.com/1204802/27448414-00cff9d4-5785-11e7-89e8-eb4fe580b2d8.png)

But I think it's okay to actually remove this advanced property for now to just get the PR merged, if that's simpler.

Alternately, I feel pretty strongly that the fixed background should be applied by default. And it should not be a `data-align` value. It should be something you can toggle _in addition_ to the image alignment.

mtias replied:
Yes, it's not an alignment, just an extra attribute.

jasmussen replied:
I gave George the go-ahead to merge this without this fixed position toggle (which we just removed). I think it's good to merge this block, and we can add the toggle separately if we desire it.

jasmussen committed:
A little polish

- Make background attachment fixed by default. Let's either remove the option to pin it entirely, or move it to the inspector and make it something that's on by default.
- Added gray scrim, removed text shadow, reduced font size
- Removed alignments because these are now inherited

Future improvements we should also consider: add the full-wide button that images have now, and make it so the dark scrim is only present when there's text.

Commit:
Re-introduce post-content data

Commit:
Add cover-image back into list

Commit:
Removing "fixed" alignment

jasmussen committed:
Fix full bleed issues.

jasmussen commented:
Okay I think this is good to go! 👍 👍

StaggerLeee commented:
May I suggest you make Caption field (Cover Image Title) a bit different than rest of the image ?
To make it clear it is a editable element inside image. Right now only visual help is a mouse cursor.

![](https://s3.postimg.org/3ontv41eb/2017-08-03_151828.jpg)
![](https://s4.postimg.org/w4j59w7x9/2017-08-03_151849.jpg)
